crates/hir_expand/src/db.rs

```diff
@@ -313,30 +342,44 @@ pub fn parse_macro_with_arg(
                 log::warn!("fail on macro_parse: (reason: {})", err);
             }
         }
+    }
+    let tt = match result.value {
+        Some(tt) => tt,
+        None => return result.drop_value(),
     };
-    let tt = tt?;
 
     let fragment_kind = to_fragment_kind(db, macro_call_id);
 
-    let (parse, rev_token_map) = mbe::token_tree_to_syntax_node(&tt, fragment_kind).ok()?;
+    let (parse, rev_token_map) = match mbe::token_tree_to_syntax_node(&tt, fragment_kind) {
+        Ok(it) => it,
+        Err(err) => {
+            return MacroResult::error(format!("{:?}", err));
+        }
+    };
 
-    if err.is_none() {
-        Some((parse, Arc::new(rev_token_map)))
-    } else {
-        // FIXME:
-        // In future, we should propagate the actual error with recovery information
-        // instead of ignore the error here.
-
-        // Safe check for recurisve identity macro
-        let node = parse.syntax_node();
-        let file: HirFileId = macro_file.into();
-        let call_node = file.call_node(db)?;
-
-        if !diff(&node, &call_node.value).is_empty() {
-            Some((parse, Arc::new(rev_token_map)))
-        } else {
-            None
+    match result.error {
+        Some(error) => {
+            // FIXME:
+            // In future, we should propagate the actual error with recovery information
+            // instead of ignore the error here.
+
+            // Safe check for recurisve identity macro
+            let node = parse.syntax_node();
+            let file: HirFileId = macro_file.into();
+            let call_node = match file.call_node(db) {
+                Some(it) => it,
+                None => {
+                    return MacroResult::error(error);
+                }
+            };
+
+            if !diff(&node, &call_node.value).is_empty() {
+                MacroResult { value: Some((parse, Arc::new(rev_token_map))), error: None }
```

Comment:
why not propagate the error here?

Reply:
ah, good idea; I wasn't fully understanding what this code tried to do, so I just matched its behavior